### PR TITLE
Wipdcf

### PR DESCRIPTION
**Pull request by ahsanjav**
- decoupled the kernel width and oversampling factor for iterative DCF estimation. 

- also fixed a bug that enables atomic convolution for estimation of DCF. 

- Fixed bugs in GIRF correction. Matched with ACW's implementation.

- bug fix in the forward NFFT that was working incorrectly for #coils > max_coils = 16. 

- Added cuda compute architectures to support compute 8+

- Added cuda fcns to support our reconstructions will push reconstructions as a separate pull-request.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (dchansen) [inline .vscode/settings.json:1]: Please remove your .vscode files from the repo. Potentially add .vscode to .gitignore

File: toolboxes/mri/pmri/gpu/cuNonCartesianSenseOperator.cu [hunk 2]
```diff
@@ -48,11 +50,36 @@ cuNonCartesianSenseOperator<REAL,D>::mult_MH( cuNDArray< complext<REAL> >* in, c
 	  throw std::runtime_error("cuNonCartesianSenseOperator::mult_MH: input/output arrays do not match specified domain/codomains");
   }
   std::vector<size_t> tmp_dimensions = *this->get_domain_dimensions();
+  std::vector<size_t> tmp_dimensions_data = *this->get_codomain_dimensions();
+  tmp_dimensions_data.pop_back(); // Remove channel dimension 
+   if(tmp_dimensions_data[tmp_dimensions_data.size()-1]==this->ncoils_ && tmp_dimensions_data.size()>2)
+     tmp_dimensions_data.pop_back(); // High risk
```
Comment (dchansen) [inline toolboxes/mri/pmri/gpu/cuNonCartesianSenseOperator.cu:56]: What are you doing here? I prefer my recons less risky - one could say "SENSEible" ;)

File: toolboxes/mri/pmri/gpu/cuNonCartesianSenseOperator.cu [hunk 3]
```diff
@@ -48,11 +50,36 @@ cuNonCartesianSenseOperator<REAL,D>::mult_MH( cuNDArray< complext<REAL> >* in, c
...
+          plan_->compute(temp_ch_data, temp_ch_recon, dcw_.get(), NFFT_comp_mode::BACKWARDS_NC2C);
+        //nfft_plan_->compute(data, *result, dcw, NFFT_comp_mode::BACKWARDS_NC2C);
+        cudaMemcpy(tmp.get_data_ptr() + tmp_dimensions[0] * tmp_dimensions[1] * tmp_dimensions[2] * iCHA,
+                   temp_ch_recon.get_data_ptr(),
+                   tmp_dimensions[0] * tmp_dimensions[1] * tmp_dimensions[2] * sizeof(complext<REAL>), cudaMemcpyDefault);
+    }
+
  // Do the NFFT
-  plan_->compute( *in, tmp, dcw_.get(), NFFT_comp_mode::BACKWARDS_NC2C );
+  //plan_->compute( *in, tmp, dcw_.get(), NFFT_comp_mode::BACKWARDS_NC2C );
```
Comment (dchansen) [inline toolboxes/mri/pmri/gpu/cuNonCartesianSenseOperator.cu:82]: Delete, or keep. There is no commented code.

File: toolboxes/mri/sdc/SDC.h [hunk 4]
```diff
@@ -4,47 +4,42 @@
...
-     * \param num_iterations Number of iterations.
-     * \return std::shared_ptr<ARRAY<REAL>> Density compensation weights, same size as trajectory.
-     */
-    template<template<class> class ARRAY, class REAL, unsigned int D>
-    std::shared_ptr<ARRAY<REAL>> estimate_dcw(
-        const ARRAY<vector_td<REAL, D>>& traj,
-        const vector_td<size_t, D>& matrix_size,
-        REAL os_factor = 2.1,
-        unsigned int num_iterations = 10);
+//#include "mri_sdc_export.h"
```
Comment (dchansen) [inline toolboxes/mri/sdc/SDC.h:7]: Delete

File: toolboxes/nfft/gpu/ConvolverNC2C_atomic.cuh [hunk 5]
```diff
@@ -89,7 +89,15 @@ static void NFFT_iterate_body(vector_td<unsigned int, D> matrix_size_os,
             (batch*num_frames+frame)*prod(matrix_size_os) + co_to_idx( vector_td<unsigned int, D>(grid_position), matrix_size_os );
 
         // Atomic update.
-        atomicAdd(&(image[grid_idx]), weight*sample_value);
+        if constexpr (is_complex_type_v<T>)
+        {
```
Comment (dchansen) [inline toolboxes/nfft/gpu/ConvolverNC2C_atomic.cuh:93]: What's wrong with the atomicAdd we already have? Also, please don't use << to divide by two. It's hacky, hard to read and (worst of all) slow.

**Review by dchansen — COMMENTED** [review]
Most of the changes look good. A few minor issues, and some details are needed for what you're doing with SENSE

File: .vscode/settings.json (hunk 1 above)
Comment (ahsanjav) [reply .vscode/settings.json:1]: Sounds good. I will remove this.

File: toolboxes/mri/pmri/gpu/cuNonCartesianSenseOperator.cu (hunk 2 above)
Comment (ahsanjav) [reply toolboxes/mri/pmri/gpu/cuNonCartesianSenseOperator.cu:56]: Hmmmm - This was my attempt to do coil by coil gridding which is significantly faster for larger data sets. I can keep it local for now or we can discuss ways to implement it in a more sensible way :) .

File: toolboxes/nfft/gpu/ConvolverNC2C_atomic.cuh (hunk 5 above)
Comment (ahsanjav) [reply toolboxes/nfft/gpu/ConvolverNC2C_atomic.cuh:93]: No issues - I just committed the bug fix from December but didn't realize you have already taken care of it as well ! This looks great - much more elegant.

**Review by dchansen — APPROVED** [review]

**Comment by hansenms** [comment]
Should this remain open? If so, please merge in master and then we will run the pipeline again.

**Comment by ahsanjav** [comment]
@dchansen : This is ready to be merged.

File: CMakeLists.txt [hunk 6]
```diff
@@ -251,7 +251,6 @@ if (CUDA_FOUND AND USE_CUDA)
         set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} "-gencode arch=compute_70,code=sm_70")
         set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} "-gencode arch=compute_75,code=sm_75")
         set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} "-gencode arch=compute_80,code=sm_80")
-        set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} "-gencode arch=compute_86,code=sm_86")
```
Comment (dchansen) [inline CMakeLists.txt:254]: Why did you remove this line?

File: gadgets/pmri/gpuCgSenseGadget.cpp [hunk 7]
```diff
@@ -153,6 +153,9 @@ namespace Gadgetron{
     std::vector<size_t> image_dims = to_std_vector(matrix_size_);
     image_dims.push_back(frames);
     
+    GDEBUG("Matrix size codomain : [%d] \n", device_samples->get_dimensions().get()->size());
```
Comment (dchansen) [inline gadgets/pmri/gpuCgSenseGadget.cpp:156]: Do we want this printed out all the time?

File: toolboxes/mri/pmri/gpu/cuNonCartesianSenseOperator.cu [hunk 8]
```diff
@@ -48,11 +50,47 @@ cuNonCartesianSenseOperator<REAL,D>::mult_MH( cuNDArray< complext<REAL> >* in, c
...
+   if(tmp_dimensions_data[tmp_dimensions_data.size()-1]==this->ncoils_ && tmp_dimensions_data.size()>2)
+   {
+     tmp_dimensions_data.pop_back(); 
+     cuNDArray<complext<REAL>> temp_ch_data(&tmp_dimensions_data);
+     tmp_dimensions.pop_back();
+     cuNDArray<complext<REAL>> temp_ch_recon(&tmp_dimensions);
+
+
+      for (int iCHA = 0; iCHA < CHA; iCHA++)
+        {
```
Comment (dchansen) [inline toolboxes/mri/pmri/gpu/cuNonCartesianSenseOperator.cu:73]: This needs a bit more explanation. And you need to verify that it is also faster for 2D, and not just 3D

File: toolboxes/mri_core/mri_core_girf_correction.cpp [hunk 9]
```diff
@@ -1,160 +1,282 @@
...
-//            size_t nlines_new = size_t(nlines * girf_sampling_time / gradient_sampling_time);
-            size_t nlines_new = girf_kernel.get_size(0);
+            using namespace std;
 
+            size_t nlines = gradients.get_size(0);
+            size_t nlines_new = girf_kernel.get_size(0) * std::round(girf_sampling_time / gradient_sampling_time); //std::max(girf_kernel.get_size(0) * girf_sampling_time / gradient_sampling_time,nlines * std::round(girf_sampling_time / gradient_sampling_time)); // This logic here may need to be checked
             size_t nbatches = gradients.get_size(1);
 
-            hoNDArray<vector_td<float,D>> result(gradients.get_dimensions());
+            //auto padded_kernel=GIRF::zeropadding(girf_kernel,int(girf_sampling_time/gradient_sampling_time));
```
Comment (dchansen) [inline toolboxes/mri_core/mri_core_girf_correction.cpp:40]: No commented code please.

File: toolboxes/mri_core/mri_core_girf_correction.cpp [hunk 10]
```diff
@@ -1,160 +1,282 @@
...
                         rotated(i, k) = gradient_vector[k];
                 }
+                float maxTx;
+                float minTx;
+                auto temp = real(permute(rotated, {0, 1}));
+                Gadgetron::maxValue(hoNDArray<float>(temp(slice, 0)), maxTx);
+                Gadgetron::minValue(hoNDArray<float>(temp(slice, 0)), minTx);
+
+                if (maxTx > 100 || minTx < -100)
+                    GERROR("What the heck");
```
Comment (dchansen) [inline toolboxes/mri_core/mri_core_girf_correction.cpp:76]: Not useful. Feels like a debug comment. If it is actually a failstate, throw an exception.

File: toolboxes/mri_core/mri_core_girf_correction.cpp [hunk 11]
```diff
@@ -1,160 +1,282 @@
...
+                // Add filter to smooth out the transition for the gradients after zero-padding
+                // This reduces ringing !
+                for (size_t k = 0; k < 3; k++)
+                {
+                    for (size_t i = nlines; i < nlines + filter.size() / 2; i++)
+                    {
```
Comment (dchansen) [inline toolboxes/mri_core/mri_core_girf_correction.cpp:82]: Please move all the filtering to its own separate function

File: toolboxes/mri_core/mri_core_girf_correction.cpp [hunk 12]
```diff
@@ -1,160 +1,282 @@
...
+                Gadgetron::maxValue(hoNDArray<float>(temp(slice, 0)), maxTx);
+                Gadgetron::minValue(hoNDArray<float>(temp(slice, 0)), minTx);
 
+                if (maxTx > 100 || minTx < -100)
+                    GERROR("What the heck");
 
-//            hoNDFFT<float>::instance()->ifft(&rotated,0);
+                for (size_t k = 0; k < 3; k++)
+                {
+
```
Comment (dchansen) [inline toolboxes/mri_core/mri_core_girf_correction.cpp:134]: What exactly are you trying to do here?

File: toolboxes/mri_core/mri_core_girf_correction.cpp [hunk 13]
```diff
@@ -1,160 +1,282 @@
...
-    }
+        hoNDArray<floatd2>
+        girf_correct(const hoNDArray<floatd2> &gradients, const hoNDArray<std::complex<float>> &girf_kernel,
+                                          const arma::fmat33 &rotation_matrix, float gradient_sampling_time, float girf_sampling_time,
+                                          float TE)
+        {
+            return girf_correct_D(gradients, girf_kernel, rotation_matrix, gradient_sampling_time, girf_sampling_time, TE);
+        }
 
+        hoNDArray<std::complex<float>> zeropadding(hoNDArray<std::complex<float>> input, int zpadFactor)
```
Comment (dchansen) [inline toolboxes/mri_core/mri_core_girf_correction.cpp:211]: What's wrong with the padding function we have? If you need a highly specific behaviour, that's fine, but keep it internal to the GIRF module.

File: toolboxes/nfft/ConvolutionKernel.hpp [hunk 14]
```diff
@@ -305,40 +305,44 @@ namespace Gadgetron
...
-        // all axes anyway.
-        this->radius_ = this->norm_radius * max(this->os_factor_);
-        this->width_ = this->radius_ * REAL(2);
-    }
+    // template<class REAL, unsigned int D>
+    // JincKernel<REAL, D>::JincKernel(
+    //     float kernelWidth)
+    //   : ConvolutionKernel<REAL, D, JincKernel>(0.0)
+    // {
+    //     // this->matrix_size_ = matrix_size;
```
Comment (dchansen) [inline toolboxes/nfft/ConvolutionKernel.hpp:335]: Delete old code

**Comment by dchansen** [comment]
It seems like you broke a test (fs-csi.xml). That would need to be fixed as well.

File: CMakeLists.txt (hunk 6 above)
Comment (ahsanjav) [reply CMakeLists.txt:254]: Added it back.

File: toolboxes/mri_core/mri_core_girf_correction.cpp (hunk 10 above)
Comment (ahsanjav) [reply toolboxes/mri_core/mri_core_girf_correction.cpp:76]: Removed.

File: toolboxes/mri_core/mri_core_girf_correction.cpp (hunk 12 above)
Comment (ahsanjav) [reply toolboxes/mri_core/mri_core_girf_correction.cpp:134]: There is some magical filtration that is needed to make GIRF work - Its just applying a filter to smooth the zero padding for the gradient waveform and rotating the waveform to the center before cetered fft. I spent some time trying to tweak this and settled on this implementation - there maybe a better way to do this but we know this works and it replicates the matlab implementation.

**Comment by hansenms** [comment]
Sorry for the build delay on this. The build agent image needed an update.

**Comment by hansenms** [comment]
@ahsanjav is this still something you want in? Can you resolve the merge conflict and address the comments form @dchansen

**Comment by hansenms** [comment]
@ahsanjav last call, then I think we close this and can reopen later if needed.

**Comment by ahsanjav** [comment]
We can revisit this when I return from my parental leave - this was not passing because the test had a bug that I believe David resolved. 

However, pulling the latest branch would also pull changes to cuda FFT which significantly slows our reconstructions and I am not sure what's the best way to proceed - perhaps we can discuss this.

**Comment by hansenms** [comment]
Let's pick it up when you are back.
